Enforce ordering of functions in synctriggers payload

balag0 opened the pull request:
Background synctriggers task compares hashes of current and last known synctriggers payload to avoid unnecessary /synctriggers call to update the site metadata. Currently the ordering of functions in Synctriggers payload is non deterministic since it relies EnumerateDirectories() which is implemented differently on Linux and Windows. This leads to unnecessary updates to the site metadata if the function app has more than 1 function even though the triggers themselves haven't been updated. Fix is to sort the functions before calculating the hash.

Fixes https://github.com/Azure/azure-functions-host/issues/6686

<!-- Please provide all the information below.  -->

### Issue describing the changes in this PR

resolves https://github.com/Azure/azure-functions-host/issues/6686

### Pull request checklist

* [X ] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ X] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [X ] Otherwise: Backport tracked by issue/PR https://github.com/Azure/azure-functions-host/issues/6686
* [X ] I have added all required tests (Unit tests, E2E tests)

<!-- Optional: delete if not applicable  -->
### Additional information

Additional PR information

balag0 commented on src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs:
```diff
@@ -267,7 +267,7 @@ internal async Task<CloudBlockBlob> GetHashBlobAsync()
         public async Task<SyncTriggersPayload> GetSyncTriggersPayload()
```
Note: Since GetSyncTriggersPayload() doesn't expose the new sortFunctionsMetadata param, triggers will be sorted for both background synctriggers and the explicit api call even though the background task is the one the issue was originally opened for.

mathewc commented on src/WebJobs.Script/Extensions/FunctionMetadataNameComparer.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace Microsoft.Azure.WebJobs.Script.Extensions
+{
+    public class FunctionMetadataNameComparer : IComparer<FunctionMetadata>
```
Recommend you just remove this class and simply sort via `functionMetadataList.OrderBy(p => p.Name).ToImmutableArray();`

balag0 committed:
Enforce ordering of functions in synctriggers payload

Background synctriggers task compares hashes of current and last known synctriggers payload to avoid unnecessary /synctriggers call to update the site metadata. Currently the ordering of functions in Synctriggers payload is non deterministic since it relies EnumerateDirectories() which is implemented differently on Linux and Windows. This leads to unnecessary updates to the site metadata if the function app has more than 1 function even though the triggers themselves haven't been updated. Fix is to sort the functions before calculating the hash.
